pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -117,6 +120,13 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		if !deployutil.IsTerminatedDeployment(latestDeployment) {
 			return nil
 		}
+
+		// As the deployment configuration has changed, we need to make sure to clean up old
+		// ones if we have now reached our deployment history quota
+		if err := c.cleanupOldDeployments(existingDeployments, config); err != nil {
```

Comment:
You should run this inside `reconcileDeployments` just before updating the status of the deploymentconfig. `reconcileDeployments` may scale an older deployment down that may be a candidate for cleaning up.